macros/edr/dbt_artifacts/upload_artifacts_to_table.sql

```diff
@@ -18,4 +18,48 @@
     {%- if should_commit -%}
         {% do adapter.commit() %}
     {%- endif -%}
+{% endmacro %}
+
+{% macro upload_csv_artifacts_to_table(table_relation, artifacts, flatten_artifact_callback) %}
+    {% set output_csv_path = elementary.get_target_path(table_relation.identifier ~ '.csv') %}
+    {% set flatten_artifact_dicts = [] %}
+    {% for artifact in artifacts %}
+        {% set flatten_artifact_dict = flatten_artifact_callback(artifact) %}
+        {% set flatten_artifact_rendered_dict = {} %}
+        {% for key in flatten_artifact_dict %}
+          {% set value = flatten_artifact_dict[key] %}
+          {% if value %}
+            {% do flatten_artifact_rendered_dict.update({key: elementary.render_csv_value(value)}) %}
+          {% endif %}
+        {% endfor %}
+        {% do flatten_artifact_dicts.append(flatten_artifact_rendered_dict) %}
+    {% endfor %}
+    {% set flatten_artifacts_agate = elementary.get_agate_table().from_object(flatten_artifact_dicts) %}
+    {% if output_csv_path.exists() %}
+      {% set cached_artifacts_agate = elementary.get_agate_table().from_csv(output_csv_path) %}
+      {% if elementary.are_artifacts_equal(cached_artifacts_agate, flatten_artifacts_agate) %}
+        {% do elementary.debug_log("[%s] Artifacts were not changed. Skipping upload." % table_relation.identifier) %}
+        {% do return(none) %}
+      {% endif %}
+    {% endif %}
+    {% do flatten_artifacts_agate.to_csv(output_csv_path) %}
+    {% do elementary.debug_log("[%s] Uploading artifacts." % table_relation.identifier) %}
+    {% do elementary.seed_elementary_model(table_relation, output_csv_path) %}
+{% endmacro %}
+
+{%- macro render_csv_value(value) -%}
+  {% if not value %}
+    {{ return(none) }}
+  {% endif %}
+
+  {% if value is not string and value is not number %}
+    {{ return(tojson(value)) }}
+  {% else %}
+    {{ return(value) }}
+  {% endif %}
+{% endmacro %}
+
+{% macro are_artifacts_equal(first_artifacts_agate, second_artifacts_agate) %}
```

Comment:
Not yet sure how performant is that.

The problem is that the artifacts contain a timestamp (`generated_at`).
This is guaranteed to change and therefore we cannot compare the file hashes which have to contain that field.

Reply:
Not the most accurate science but seems to be fast.

Created sample data with 25,000 rows where each row is 500 characters:

```py
    with open('raw_data.csv', 'w') as f:
    writer = csv.writer(f)
    writer.writerow(['hello', 'world', 'test'])
    for _ in range(25000):
        writer.writerow([os.urandom(100), os.urandom(200), os.urandom(300)])
```

Both loading the table and and comparing the rows (`t1.rows == t2.rows`) were very fast.